### DISCUSSION TITLE
Requests In Nextjs

### DISCUSSION THREAD

**Discussion by FarNys** [vercel/next.js#42946] [Help] [answered]
Hi everyone,
Lately i built a fullstack app with Nextjs, backed with Strapi. it work well in dev mode.
but there is some weird problem in production mode.
all my requests in getStaticProps And getServerSideProps works well, but same requests in useEffect(Client Reqs) give some error (all request type like get,post ,..), i check error and i found it changes all http reqs to https(http://192.....:1337/api to https://192.....:1337/api) and so it gives SSL_Error.
(Both frontend and backend are in same server with different ports. there is SSL for frontend)
Is there any suggestion?:)

**Comment by icyJoseph** [comment, marked as answer]
To debug, or realise what's going on here, you could drop the front end, grab POSTMAN or just from the browser console try to reach for your endpoint. 

Most likely the backend server is not reachable via HTTPS protocol, only via HTTP. Servers require a bit of setup to get them to open up under HTTPS. Read up a bit on how to set that up. Alternatively try Caddy reverse proxy to expose your server via HTTPS.

- https://caddyserver.com/docs/quick-starts/reverse-proxy#https-from-client-to-proxy

**Reply by FarNys** [reply]
Thanks _icyJoseph_  for answer.

i found the problem, because my site has ssl (https://example.com), browsers automatically change all http requests-> https (CSP-https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP).
after some struggle i found strapi will change url in domain, and its api url is difference with ip vs domain.

(i added url:https://example.com/panel in config/server.js as my base url for strapi admin and config my Nginx based on that, but it wont affect local server api request but will change request through domain address).

example: **http://10.10.10.10:1337/api** (test with local server and got respose) --> **https://example.com/panel/api**

**Reply by icyJoseph** [reply]
Aha, so you have the `upgrade-insecure-requests` CSP flag set, correct?

I'd say, it might be better, perhaps not ideal, to proxy the requests to your strapi API.

Just use `next.config.js` rewrites, and from the client call `/api/strapi` (name's up to you), and in rewrites:

```js
module.exports = {
  rewrites() {
    return [
      {
        source: "/api/strapi",
        destination: "https://example.com/panel/api",
      },
    ];
  },
};
```

Something like that :) then your client should just fetch from `/api/strapi`.

Alternatively, something more complex like: https://maxschmitt.me/posts/next-js-api-proxy/

**Reply by FarNys** [reply]
I wont set upgrade-insecure-requests flag, just simple fetch request without any extra header.
but it seems modern browsers will add this header automatically.

whats the point of proxy request in next.config.js?

can we do something like this?( just for example:) )
```
module.exports = {
        rewrites() {
            return [
                {
                    source: "https://example.com",
                    destination: "http"://example.com,
             },
        ];
    },
};
```

**Reply by icyJoseph** [reply]
Ah no, wait I may have made an assumption... Let's go step by step

How are you connecting to your Strapi from the browser? Are you using some SDK, or just simple `fetch`?

> And no, you can only rewrite incoming requests coming to your server, AKA relative paths if that makes sense to you

**Reply by FarNys** [reply]
i mean i wrote simple fetch request in my useEffect(Nextjs) and deploy in production.

if you write http request in your client side app(i mean not in getStaticProps or ServerSideProps), when you deploy it and add SSL to it, you cant reach your endpoint via your app because browser will change your http to https.

as i said before my problem solved with changing request url like http:10.10.10.10.:1337/api to https://example.com/panel/api.

**Reply by icyJoseph** [reply]
Aha, I did not understand you had solved it :)!

My suggestion was to change `fetch("http:10.10.10.10.:1337/api")` to `fetch("/api/strapi")`, and then tell Next.js to rewrite `/api/strapi` to `https://example.com/panel/api` so that the client doesn't have to know about the actual URL where the Strapi API lives.

`next.config.js` has access to the env vars, so you can make use an env var to control the redirection.

Regarding:

>  browser will change your http to https.

Yeah I think, they have an upgrade when possible mechanism.

Don't forget to mark as answered.
